include/circt/Dialect/FIRRTL/FIRRTLConnectionGraph.h

```diff
@@ -0,0 +1,162 @@
+//===- FIRRTLConnectionGraph.h - Graph of connections in FIRRTL -*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides utilities for working with FIRRTL operations using LLVM
+// graph utilties.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_FIRRTL_FIRRTLCONNECTIONGRAPH_H
+#define CIRCT_DIALECT_FIRRTL_FIRRTLCONNECTIONGRAPH_H
+
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/Operation.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/ADT/iterator.h"
+#include <mlir/IR/Value.h>
+#include <mlir/IR/ValueRange.h>
+#include <variant>
+
+namespace circt {
+namespace firrtl {
+namespace detail {
+
+// "Using" is used here to avoid polluting the global namespace with
+// CIRCT-specific graph traits.  This pattern is borrowed from HWModuleGraph.h.
+using FIRRTLOperation = mlir::Operation;
```

Comment:
I'm not too familiar with graph traits, and maybe I am misunderstanding the point, but I'm not sure this is achieving anything.

Reply:
Maybe you can just add a wrapper struct and specialize on that? It could be nice way to help distinguish between the root module and the ops beneath it we are actually walking :) ?

```c++
namespace circt::firrtl {
struct ConnectionGraph {
  explicit ConnectionGraph(FModuleLike *root) : root(root) {}
  FModuleLike *root;
};
} // namespace circt::firrtl
```

```c++
namespace llvm {
template <>
struct GraphTraits<circt::firrtl::ConnectionGraph> {
  using ChildIteratorType = circt::firrtl::detail::ConnectionIterator;
  using Node = mlir::Operation;
  using NodeRef = Node *;

  static NodeRef getEntryNode(NodeRef op) { return op; }
  // NOLINTNEXTLINE(readability-identifier-naming)
  static ChildIteratorType child_begin(NodeRef op) {
    return circt::firrtl::detail::ConnectionIterator::childBegin(op);
  }
  
  // NOLINTNEXTLINE(readability-identifier-naming)
  static ChildIteratorType child_end(NodeRef op) {
    return circt::firrtl::detail::ConnectionIterator::childEnd(op);
  }
};
} // namespace llvm
```